### PR TITLE
Added Support for SK6812-Strips

### PR DESCRIPTION
**Pull request by Rdlprmpf**
Moved existing sample to subfolder, added new sample for SK6812, updated the Readme.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/iot/pull/1709)

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/iot?pullRequest=1709) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/iot?pullRequest=1709) before we can accept your contribution.<br/><br/>:x: Rdlprmpf [sign now](https://cla.dotnetfoundation.org/dotnet/iot?pullRequest=1709)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/iot?pullRequest=1709) it.</sub>

File: src/devices/Ws28xx/BitmapImageNeo4.cs [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+            Data[offset++] = _lookup[c.R * BytesPerComponent + 2];
+            Data[offset++] = _lookup[c.B * BytesPerComponent + 0];
+            Data[offset++] = _lookup[c.B * BytesPerComponent + 1];
+            Data[offset++] = _lookup[c.B * BytesPerComponent + 2];
+            Data[offset++] = _lookup[c.A * BytesPerComponent + 0];
+            Data[offset++] = _lookup[c.A * BytesPerComponent + 1];
+            Data[offset++] = _lookup[c.A * BytesPerComponent + 2];
+        }
+
+        private static readonly byte[] _lookup = new byte[256 * BytesPerComponent];
```
Comment (Ellerbach) [inline src/devices/Ws28xx/BitmapImageNeo4.cs:43]: As you're using 2 times the 256, maybe set it up as a private constant like for the others. Also a little comment of with 256 will help for the maintenance.

File: src/devices/Ws28xx/README.md [hunk 2]
```diff
@@ -66,7 +67,8 @@ void Rainbow(Ws28xx neo, int count, int iterations = 1)
     }
 }
 ```
-
+#### Note: 
+Using the SK6812 is almost the same, but the alpha channel of the color is used for the white LED. This means that the predefined color definitions (like ```System.Drawing.Color.Red``` will not work correctly as they have the alpha channel set to 255 (0xFF). That will turn the white LED always on. See the [sample](samples/SK6812_Samples/Programs.cs) for the main differences to the above code.
```
Comment (Ellerbach) [inline src/devices/Ws28xx/README.md:71]: please add an extra enter after and before the title. In Markdown, each title has to have an extra enter before and after it.

Comment (Ellerbach) [reply src/devices/Ws28xx/README.md:71]: To be more explicit as in the sample it's not that obvious, can you add a code snippet here to show how to handle the color management?

File: src/devices/Ws28xx/samples/Sk6812_Samples/Program.cs [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+using Iot.Device.Graphics;
+using Iot.Device.Ws28xx;
+
+// Configure the count of pixels
+const int Count = 8;
+Console.Clear();
+
+SpiConnectionSettings settings = new(0, 0)
+{
+    ClockFrequency = 2_400_000,
```
Comment (Ellerbach) [inline src/devices/Ws28xx/samples/Sk6812_Samples/Program.cs:18]: Please add comments in this sample to explain that this frequency is mandatory to reflect the comments in the code.

**Review by Ellerbach — APPROVED** [review]
Thanks a lot! Looks good, few adjustments to make it even easier for developers to use.

**Comment by Rdlprmpf** [comment]
I've made the requested changes. I hope the comments make it somewhat more clear. If there are any more change requests I'd be happy to work further on this.

For example I'm thinking about a color-extension-class which allows the easy "transformation" of `System.Drawing.Color` to some kind of `LedColor`.

File: src/devices/Ws28xx/samples/Sk6812_Samples/Program.cs (hunk 3 above)
Comment (krwq) [reply src/devices/Ws28xx/samples/Sk6812_Samples/Program.cs:20]: if it's mandatory consider adding DefaultSpiClockFrequency public constant to the binding and using it in the sample

Comment (Rdlprmpf) [reply src/devices/Ws28xx/samples/Sk6812_Samples/Program.cs]: > if it's mandatory consider adding DefaultSpiClockFrequency public constant to the binding and using it in the sample

Good point, done so.

File: src/devices/Ws28xx/BitmapImageNeo4.cs [hunk 4]
```diff
@@ -0,0 +1,60 @@
+﻿using System.Drawing;
```
Comment (krwq) [inline src/devices/Ws28xx/BitmapImageNeo4.cs:1]: copyright header

File: src/devices/Ws28xx/README.md [hunk 5]
```diff
@@ -1,19 +1,20 @@
-﻿# Ws28xx LED drivers
+﻿﻿# Ws28xx/SK7812 LED drivers
```
Comment (krwq) [inline src/devices/Ws28xx/README.md:1]: 6812?

Comment (krwq) [reply src/devices/Ws28xx/README.md:1]: also add spacing around `/` (here and below)

File: src/devices/Ws28xx/samples/Sk6812_Samples/Program.cs [hunk 6]
```diff
@@ -0,0 +1,152 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
```
Comment (krwq) [inline src/devices/Ws28xx/samples/Sk6812_Samples/Program.cs:4]: I'd suggest to improve original sample and perhaps add a menu in there so you can choose which LED strip you have.

Also we should file an issue about renaming base class and changing namespace to something like `LedStrip` - do not do it in this PR though

**Comment by joperezr** [comment]
[Triage] @Rdlprmpf looks like this PR is very close to be ready to merge, just waiting on resolving the conflicts and addressing the small comments that @krwq added. Do you have time to take care of that so we can go ahead and merge?

File: src/devices/Ws28xx/samples/Sk6812_Samples/Program.cs (hunk 6 above)
Comment (Rdlprmpf) [reply src/devices/Ws28xx/samples/Sk6812_Samples/Program.cs:4]: For this pull request I'd like to keep them separated. The next thing I'm trying to implement is support for WS2815, which currently don't seem to work although they're pretty similar to WS2812b. When I've got that working, I'd also rework the samples.

**Comment by joperezr** [comment]
@Rdlprmpf thanks for rebasing and fixing the merge conflicts. Looks like you have some markdown lint errors that we need to fix on some of the modified markdown files, could you please take a look at that?

```
D:/a/1/s/src/devices/Ws28xx/README.md:1 MD041/first-line-heading/first-line-h1 First line in a file should be a top-level heading [Context: "﻿# Ws28xx / SK6812 LED drivers"]
D:/a/1/s/src/devices/Ws28xx/README.md:71 MD001/heading-increment/header-increment Heading levels should only increment by one level at a time [Expected: h3; Actual: h4]
D:/a/1/s/src/devices/Ws28xx/README.md:71:11 MD009/no-trailing-spaces Trailing spaces [Expected: 0 or 2; Actual: 1]
D:/a/1/s/src/devices/Ws28xx/README.md:71:10 MD026/no-trailing-punctuation Trailing punctuation in heading [Punctuation: ':']
D:/a/1/s/src/devices/Ws28xx/README.md:75 MD031/blanks-around-fences Fenced code blocks should be surrounded by blank lines [Context: "```"]
D:/a/1/s/src/devices/Ws28xx/README.md:75 MD040/fenced-code-language Fenced code blocks should have a language specified [Context: "```"]
```

File: src/devices/Ws28xx/samples/Sk6812_Samples/Program.cs (hunk 6 above)
Comment (krwq) [reply src/devices/Ws28xx/samples/Sk6812_Samples/Program.cs:4]: fair enough

**Comment by joperezr** [comment]
Thanks so much for the contribution @Rdlprmpf!

**Comment by Rdlprmpf** [comment]
You're welcome. I'll continue to work on this as stated above. Currently I'm having a problem regarding the timings of the WS2815, but that's a different story ;-)

**Pull request merged**